packages/storage/src/tools.ts

```diff
@@ -17,3 +17,81 @@ export const addClearMethod = <S extends Storage | StorageWithOptions<any>>(
   };
   return storage as S;
 };
+
+const methodKeys = ["clear", "getItem", "getAll", "setItem", "removeItem", "key", "getLength"];
+
+/**
+ * adds a `.withOptions` method that wraps the storage to apply options
+ */
+export const addWithOptionsMethod = <
+  O,
+  S extends StorageWithOptions<O> | AsyncStorageWithOptions<O>,
+  W extends AsyncStorage | Storage = S extends AsyncStorageWithOptions<O> ? AsyncStorage : Storage,
+>(
+  storage: S,
+): S & { withOptions: (options: O) => W } => {
+  storage.withOptions = (options: O): W =>
+    methodKeys.reduce(
+      (wrapped: Partial<S>, key: keyof S) => {
+        if (typeof storage[key] === "function") {
+          (wrapped as any)[key] = (...args: Parameters<(typeof storage)[typeof key]>) =>
+            storage[key]([...args, options]);
+        }
+        return wrapped;
+      },
+      {
+        get length() {
+          return storage.length;
+        },
+        withOptions: (options: O) => storage.withOptions!(options),
+      } as Partial<S>,
+    ) as unknown as W;
+  return storage as S & { withOptions: (options: O) => W };
+};
+
+const callStorageMethod = (
+  storages: (Storage | AsyncStorage)[],
+  method: string,
+  args: any[] = [],
+  first = false,
```

Comment:
I'm not sure if Promise.race is beneficial for debugging. Because if one storage is faulty it might result in errors that are not consistantly reproducable. I think we should add an option that enables this and mention it in the documenation that this is an optimization, but the developer should be aware of what happens in the background

Reply:
But what would be the other option? We cannot infer which API is faulty and which is not, even if we wait for all of the results, which is why I stress in the README that you have to await for writing access in order to avoid race conditions.

Reply:
hmm okay i guess, i fine with that